### ISSUE TITLE
Highlight authorship in a color independent way

### ISSUE THREAD

**Issue by nidico** [codimd/react-client#610]
[accessibility]

## Problem:

Visually impaired users might struggle trying to visually differentiate user authorship, as authorship information is currently only shown by using individually colored underlines by user.

As soon as there are multiple users, the contrast between different users' colors will be very low and hard to recognized for people without full color vision.

## Solution:

Additionally or alternatively display authorship information in a color-independent way.

**Proposal 1**

In addition to assigning colors, assign different underline styles to users (manually or automatically).

_Caveat:_ It's hard to come up with many different underline styles, e.g. [css-decoration-style](https://developer.mozilla.org/en-US/docs/Web/CSS/text-decoration-style) has only 5 types.

**Proposal 2**:

Start each "underline" with a (small) number or a letter identifying the author (e.g. the first character of the username)

_Caveat:_ This might involve increasing the line spacing and should therefore be displayed optionally.

**Proposal 3**:

Combine proposal 1 and proposal 2.

## Related issues:

- codemirror/CodeMirror#4604
- https://github.com/hackmdio/codimd/issues/1020

**Comment by ccoenen** [comment]
I would not try to have this "perfect" at the expense of "impossible to implement". i.e. "done" is better than "perfect".

I'd argue that 5 distinct styles are _way_ better than none. I would add, that many/most documents probably have a small handfull of authors, so if one were to divide them by the four leading contributors (each receiving their own style) and a fifth style for any "also ran" contributors, I would assume that this will be an amazing step up from an accessibility standpoint. This fifth style should always be reserved for "means many other people" consistently.

(same if there's only 3 or 4 distinct styles we can come up with!)

**Comment by DerMolly** [comment]
As we support underline as part of the markdown spec `++like so++`I guess we only have 4 styles to choose from
- double underline ![image](https://user-images.githubusercontent.com/3978662/94359723-2975ef80-00a9-11eb-9919-d74aa1edc17f.png)
- dotted underline ![image](https://user-images.githubusercontent.com/3978662/94359751-3e528300-00a9-11eb-98e5-0729f562c110.png)
- dashed underline ![image](https://user-images.githubusercontent.com/3978662/94359764-4dd1cc00-00a9-11eb-845c-0367ffbdcc91.png)
- wave underline ![image](https://user-images.githubusercontent.com/3978662/94359782-5cb87e80-00a9-11eb-93d9-112c6b4072d2.png)

so by @ccoenen suggestion that would leave us with three styles for the leading contributors and one catch all for everyone else. (if we don't want to exclude wave underline as a way to mark spellchecker errors)

**Comment by SISheogorath** [comment]
Yep, we should definitely **not** use wave underline for authorship.

**Comment by mrdrogdrog** [comment]
IMHO the css underline options are to similar. I don't think that they would make any difference for the most people.

**Comment by mrdrogdrog** [comment]
> Yep, we should definitely **not** use wave underline for authorship.

Also: this. If I see wavy underlining then I would suggest "I misspelled this" :sweat_smile:

**Comment by ccoenen** [comment]
I am not visually impaired myself, but from the design of accessible charts and graphics I know that even slight alterations in patterns help distinguish various datasets. Also keep in mind that people that suffer from colorblindness are often compensating by picking up things like these pattern-nuances. I am not an authority on these topics, so eventually it would be great to talk to someone who is.

I am, for example, not sure if there's something that ties into this from a technological standpoint. For example if there's a more standardized variant of the [`-ms-high-contrast` media query](https://developer.mozilla.org/en-US/docs/Web/CSS/Media_Queries/Using_Media_Queries_for_Accessibility) or something like that. I also know very little about screen readers - which we eventually also should support better.

**Comment by davidmehren** [comment]
An idea I had for generating custom underlines:
- get `n` bits
- set a few of then (randomly or just count)
- generate a pattern from the set bits
- paint the pattern underneath each character

That will get you color-independent underlines for `2^n` authors. We would need to find out how big `n` can be, so that the patterns are still distinguishable. We could also try 2-lined patterns, that might improve distinguishability. See also my high-quality ASCII-art below :smile: 


```
  _                 _     
 | |_   ___  __ __ | |_   
 |  _| / -_) \ \ / |  _|  
  \__| \___| /_\_\  \__|  
 12345 12345 12345 12345 <-- one line of pattern-bits
                       _                  
  ___   _ _    ___    | |_  __ __ __  ___ 
 / _ \ | ' \  / -_)   |  _| \ V  V / / _ \
 \___/ |_||_| \___|    \__|  \_/\_/  \___/
 ██  █ ██  █  ██  █   █   █  █   █   █   █
 ^                    ^
 author1              author2
                                               

  _                 _   
 | |_   ___  __ __ | |_ 
 |  _| / -_) \ \ / |  _|
  \__| \___| /_\_\  \__|  
 12345 12345 12345 12345
 67890 67890 67890 67890 <-- two lines of pattern-bits

                       _                  
  ___   _ _    ___    | |_  __ __ __  ___ 
 / _ \ | ' \  / -_)   |  _| \ V  V / / _ \
 \___/ |_||_| \___|    \__|  \_/\_/  \___/
 ██  █ ██  █  ██  █   █   █  █   █   █   █
   █ █   █ █    █ █      ██     ██      ██
 ^                    ^
 author1              author2
```

**Comment by mrdrogdrog** [comment]
Well it would be possible. As a certified CSS-Magician I can approve, that this wouldn't be much more complicated than the normal underlining.

**Comment by ccoenen** [comment]
This is a problem that has been solved before,  maybe there's something to be learned here:
https://observablehq.com/@bradydowling/textures
Library in use: https://riccardoscalco.it/textures/
